core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -396,14 +411,15 @@ private ThreadContextStruct putResponseHeaders(Map<String, List<String>> headers
             return new ThreadContextStruct(requestHeaders, newResponseHeaders, transientHeaders);
         }
 
-        private ThreadContextStruct putResponse(String key, String value) {
+        private ThreadContextStruct putResponse(final String key, final String value, final Function<String, String> uniqueValue) {
             assert value != null;
 
             final Map<String, List<String>> newResponseHeaders = new HashMap<>(this.responseHeaders);
             final List<String> existingValues = newResponseHeaders.get(key);
 
             if (existingValues != null) {
-                if (existingValues.contains(value)) {
+                final Set<String> existingUniqueValues = existingValues.stream().map(uniqueValue).collect(Collectors.toSet());
```

Comment:
assert that existingUniqueValues have the same length as existingValues? we should use deduping consistently...